tools/jsonschemavalidator/main.cpp

```diff
@@ -0,0 +1,204 @@
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+
+=============================================================================*/
+
+#include "optionparser.h"
+#include "rapidjson/error/en.h"
+#include "rapidjson/filereadstream.h"
+#include "rapidjson/schema.h"
+#include "rapidjson/stringbuffer.h"
+#include "rapidjson/prettywriter.h"
+#include <memory>
+#include <iostream>
+#include <string>
+
+using namespace rapidjson;
+
+// functor object for deleting FILE objects
+struct FILEDeleter
+{
+  typedef FILE* pointer;
+  void operator()(FILE *pFile)
+  {
+    if (pFile)
+      fclose(pFile);
+  }
+};
+
+/**
+ * \brief Utility function to create a rapidjson::Document object from a file path
+ *
+ * \jsonfile is the file path of the json data
+ *
+ * \returns rapidjson::Document object representing the json file
+ * \throws std::runtime_error if the input file is non-existent or is not a valid json file
+ */
+Document ReadJSONDocument(const std::string& filePath)
+{
+  std::unique_ptr<FILE, FILEDeleter> fp(fopen(filePath.c_str(), "r"));
+  if (!fp)
+  {
+    std::string err("File '" + filePath + "' not found");
+    throw std::runtime_error(err);
+  }
+  char buffer[4096];
+  FileReadStream fs(fp.get(), buffer, sizeof(buffer));
+  Document doc;
+  doc.ParseStream(fs);
+  if(doc.HasParseError())
+  {
+    throw std::runtime_error("File '" + filePath + "' is not a valid JSON\n Error(offset " + std::to_string(doc.GetErrorOffset()) + "): " + GetParseError_En(doc.GetParseError()));
+  }
+  return doc;
+}
+
+/**
+ * \brief Function to validate json file against a schema file
+ *
+ * \schemafile is the file path of the schema
+ * \jsonfile is the file path of the json data
+ *
+ * \returns std::pair<bool, std::string>. The first element of the pair is true
+ *          if the json file matches the schema, error otherwise. In case of
+ *          error, the second element contains the error report
+ */
+std::pair<bool, std::string> validate(const std::string& jsonfile, const std::string& schemafile) noexcept
+{
+  try
+  {
+    SchemaDocument sd(ReadJSONDocument(schemafile));
+    SchemaValidator validator(sd);
+    Document json = ReadJSONDocument(jsonfile);
+    if (!json.Accept(validator))
+    {
+      std::string errorMsg("JSON file " + jsonfile + " does not match schema file " + schemafile);
+      StringBuffer sb;
+      PrettyWriter<StringBuffer> w(sb);
+      validator.GetError().Accept(w);
+      errorMsg.append("\nError report:\n");
```

Comment:
should the "\n" be std::endl to ensure the display is right across all platforms?